### PR TITLE
lib.network: add ipv6 parser

### PR DESCRIPTION
**Pull request by woojiq**
### Description of changes

* Add a library function to parse and validate an IPv6 address from a
string. It can parse the first two versions of an IPv6 address according
to https://datatracker.ietf.org/doc/html/rfc4291#section-2.2. The third
form "x:x:x:x:x:x.d.d.d.d" is not yet implemented.
* Add shell script network.sh to test IPv6 parser functionality.
* Copy fromHexString function from the https://github.com/NixOS/nixpkgs/pull/266705

Mentor: @Janik-Haag 

### Things done
- [x] Tested, as applicable:
* [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
* and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
* or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
- [x] Formatted using nixfmt-rfc-style

Add a 👍 [reaction](https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to [pull requests you find important](https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: lib/network/default.nix
Comment (mweinelt) [inline lib/network/default.nix:8]: Please talk about expanding and compressing, as that is the official lingo used in the RFC 4291. And make sure to reference RFC as the canonical source for these representations.

> Expand an IPv6 Address according to RFC 4291, by doing this and that

File: lib/network/default.nix
Comment (Janik-Haag) [inline lib/network/default.nix:12]:
```suggestion
    _normalizeIp =
```
This is already defined in the ipv6 scope, so I think it's preferable to have:
```nix
lib.network.ipv6._normalizeIp
```
over
```nix
lib.network.ipv6._normalizeIpv6
```

File: lib/network/default.nix
Comment (Janik-Haag) [inline lib/network/default.nix:3]: for ergonomics it would be nicer to just use v6 and v4 but I don't have hard feelings here.

File: lib/network/default.nix
Comment (Janik-Haag) [inline lib/network/default.nix:26]: I think we should keep it simple and just write "::" over introducing a extra binding, because it makes the code easier to read since you don't have to look up how doubleColon.

```suggestion
          hasPrefix = lib.strings.hasPrefix "::" addr;
          hasSuffix = lib.strings.hasSuffix "::"  addr;
          hasInfix = lib.strings.hasInfix "::" addr;
```

File: lib/network/default.nix
Comment (Janik-Haag) [inline lib/network/default.nix:23]: emptyCnt is only used once in line 38, so I think we should move it directly into the local scope. I also think that we should call this emptyCount since everything else is written out and it's only internal anyways.

File: lib/network/default.nix
Comment (Janik-Haag) [inline lib/network/default.nix:79]:
```suggestion
      builtins.map (piece: u16FromHexStr piece) addr;
```

We should keep the same language everywhere to avoid confusion.

**Comment by Janik-Haag** [comment]
To address your questions:

> Do I need to somehow hide the internal functions more than just underscoring them?

Yes currently they would all get exposed, you have to put them into a let binding to make them truly internal. Take a look at https://github.com/NixOS/nixpkgs/blob/master/lib/fileset/default.nix#L103 as an example.

> Are the doc comments in the correct format (already existing files have different style)?

Some of them are some of them aren't, please take a look at the [example](https://github.com/nix-community/nixdoc?tab=readme-ov-file#example)

> Do I need to pull the ipv4 parser implementation to merge them together? https://github.com/NixOS/nixpkgs/pull/299409

no, we should keep them separate and preferably go for small prs, e.g. once the parser is fine merge this pr and then do another v6 one for cidr related functions.

File: lib/network/default.nix
Comment (woojiq) [reply lib/network/default.nix:3]: Personally, I think the "ip" prefix gives more clarity.
`network.v6.func` looks weird.

File: lib/tests/network.sh
Comment (woojiq) [reply lib/tests/network.sh:72]: I think we can use reserved ip in examples but for testing purpose reserved ip might not be enough.

**Comment by woojiq** [comment]
@Janik-Haag @mweinelt I refactored the code and moved internal functions to internal.nix. Can you review again please?

File: lib/network/default.nix
Comment (mweinelt) [reply lib/network/default.nix:6]: I agree, proper lingo everywhere is what we should aim for.

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
I think this is getting into shape. There are some smaller points left to do, and you should add a lot more comments explaining different parts, I think. This part of lib currently is very undocumented internally compared to others like lib.fileset or lib.generators.

File: lib/network/internal.nix
Comment (woojiq) [reply lib/network/internal.nix:25]: RFC 4291:

> In order to make writing addresses containing zero bits easier, a special syntax is available to compress the zeros.

> or in compressed form:

File: lib/network/internal.nix
Comment (woojiq) [reply lib/network/internal.nix:128]: I agree, 128 sounds more logical if prefix length is omitted.

> I think we should be stricter with the user input and don't silently just set some prefix but instead fail saying that it's not a cidr address.

I added one `fromString` function and user can also bot default and cidr notation. I don't see a reason to create two separate API functions like `fromString` and `fromCidrString`, because this can be easily observed from the passed string - if it contains a `/`, then we try to parse it as cidr notation.

Comment (woojiq) [reply lib/network/internal.nix]: Ok, when I started working on ipv6 types, I must have understood why we need separate functions for default and cidr notation. In progress.

File: lib/network/default.nix
Comment (Janik-Haag) [inline lib/network/default.nix:11]: >       The form "x:x:x:x:x:x:d.d.d.d" is not yet implemented.

We probably want to be more explicit and say The form... won't be implemented.

**Review by Janik-Haag — COMMENTED** [review]
@infinisil as a lib code owner, do you want to give this a review before we merge it?

**Review by Janik-Haag — APPROVED** [review]

File: .gitignore
Comment (drupol) [reply .gitignore:10]: I would add that this shouldn't be part of this PR and added into the user global's `.gitignore` file.

Comment (Janik-Haag) [reply .gitignore:10]: I think it's reasonable to have this, esp. since we also have it for other editors like .idea, .vscode, etc... but yes I agree it would have been more transparent if we moved it into a separate pr.

**Pull request merged**
